src/HexagonOptimize.cpp

```diff
@@ -1605,6 +1606,271 @@ class OptimizeShuffles : public IRMutator {
+    Expr calc_load(const Expr &e) {
+        if (e.as<Variable>()) {
+            return calc_load(get_val(e));
+        }
+        if (const Cast *maybe_cast = e.as<Cast>()) {
+            return calc_load(maybe_cast->value);
+        }
+        if (const Shuffle *maybe_shuffle = e.as<Shuffle>()) {
+            if (maybe_shuffle->is_slice() && maybe_shuffle->slice_stride() == 1) {
+                Expr maybe_load = calc_load(maybe_shuffle->vectors[0]);
+                if (!maybe_load.defined()) {
+                    return Expr();
+                }
+                const Load *res = maybe_load.as<Load>();
+                if (const Ramp *ramp = get_val(res->index).as<Ramp>()) {
+                    if (!ramp) {
+                        return Expr();
+                    }
+                    Expr shifted_ramp = Ramp::make(ramp->base + maybe_shuffle->slice_begin(),
+                                                    ramp->stride, ramp->lanes);
+                    Expr shifted_load = Load::make(res->type, res->name, shifted_ramp,
+                                                    res->image, res->param, res->predicate);
+                    return shifted_load;
+                }
+            }
+            else if (maybe_shuffle->is_concat()) {
+                return are_contiguous_vectors(maybe_shuffle->vectors);
+            }
+        }
+        if (const Load *maybe_load = e.as<Load>()) {
+            if (const Ramp* ramp = maybe_load->index.as<Ramp>()) {
+                if (is_const(ramp->stride, 1)) {
+                    return maybe_load;
+                }
+            }
+        }
+        return Expr();
+    }
+
+    // Loads comparator for sorting Load Expr of the same buffer.
+    static bool loads_comparator(LoadIndex a, LoadIndex b) {
+        if (a.first.defined() && b.first.defined()) {
+            const Load* load_a = a.first.as<Load>();
+            const Load* load_b = b.first.as<Load>();
+            const Ramp* ramp_a = load_a->index.as<Ramp>();
+            const Ramp* ramp_b = load_b->index.as<Ramp>();
+            if (ramp_a && ramp_b && load_a->name == load_b->name) {
+                Expr base_diff = simplify(ramp_b->base - ramp_a->base);
+                if (is_positive_const(base_diff)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    // Vtmpy helps in sliding window ops of the form a*v0 + b*v1 + v2.
+    // Conditions required:
+    //      v0, v1 and v2 start indices differ by vector stride
+    // Current supported value of stride is 1.
+    // TODO: Add support for any stride.
+    void visit(const Add *op) {
+        // Find opportunities vtmpy
+        if (op->type.is_vector() && op->type.bits() == 16) {
+            int lanes = op->type.lanes();
+            vector<MulExpr> mpys;
+            Expr rest;
+            string vtmpy_suffix;
+
+            if (op->type.bits() == 16) {
```

Comment:
It looks like this if is redundant with the one a few lines up (this should always be true inside the if from line 1773).